### PR TITLE
[SYCL] Fix mem-leak/potential SegFault in LIT test, add 2 required bu…

### PR DESCRIPTION
**Pull request by v-klochkov**
…ffer ctors

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
@v-klochkov, are you still going to commit this patch?

**Comment by v-klochkov** [comment]
> @v-klochkov, are you still going to commit this patch?

This PR adds 2 unspecified constructors to buffer class. It is better to wait for those constructors to be added to SYCL-2020 SPEC before proceeding with the checkin.
I can cancel this PR if it inconvenient to have this PR here/now.

**Comment by bader** [comment]
> This PR adds 2 unspecified constructors to buffer class.
> It is better to wait for those constructors to be added to SYCL-2020 SPEC before proceeding with the checkin.

@intel/dpcpp-specification-reviewers, @gmlueck, are these constructors expected to be added to SYCL specification?

**Comment by v-klochkov** [comment]
> > This PR adds 2 unspecified constructors to buffer class.
> > It is better to wait for those constructors to be added to SYCL-2020 SPEC before proceeding with the checkin.
> 
> @intel/dpcpp-specification-reviewers, @gmlueck, are these constructors expected to be added to SYCL specification?

Added @mkinsner to reviewers, the issue was discussed with Mike previously.

**Review by bader — CHANGES_REQUESTED** [review]
Please, add a regression test.

**Comment by v-klochkov** [comment]
> Please, add a regression test.

Added in da51ec6df1858ddb489595eaafd7f9ad6f99d251

**Comment by v-klochkov** [comment]
2 buffer constructors implemented in this patch are part of SYCL-2020: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_buffer_interface

File: sycl/test/basic_tests/buffer/buffer_ctad.cpp [hunk 1]
```diff
@@ -19,22 +12,45 @@ int main() {
...
   buffer b3a(cv);
   static_assert(std::is_same<decltype(b3a), buffer<int, 1>>::value);
+
   shared_ptr_class<int> ptr{new int[5], [](int *p) { delete[] p; }};
   buffer b4(ptr, range<1>(5));
   static_assert(std::is_same<decltype(b4), buffer<int, 1>>::value);
+
+  std::allocator<int> buf_alloc;
+  shared_ptr_class<int> ptr_alloc{new int[5], [](int *p) { delete[] p; }};
+  buffer b5(ptr, range<1>(5), buf_alloc);
```
Comment (bader) [inline sycl/test/basic_tests/buffer/buffer_ctad.cpp:44]:
```suggestion
  buffer b5(ptr_alloc, range<1>(5), buf_alloc);
```

Comment (v-klochkov) [reply sycl/test/basic_tests/buffer/buffer_ctad.cpp]: Thank you. Fixed in 806566da62542c1a7955fb97558819eaebbfe42f

**Review by bader — APPROVED** [review]

**Comment by v-klochkov** [comment]
@rbegam, @intel/llvm-reviewers-runtime  - please review/approve the fix.

**Review by againull — APPROVED** [review]

**Pull request merged**